mmrotate/models/dense_heads/csl_rotated_retina_head.py

```diff
@@ -0,0 +1,564 @@
+
+        Args:
+            cls_score (torch.Tensor): Box scores for each scale level
+                Has shape (N, num_anchors * num_classes, H, W).
+            bbox_pred (torch.Tensor): Box energies / deltas for each scale
+                level with shape (N, num_anchors * 5, H, W).
+            anchors (torch.Tensor): Box reference for each scale level with
+                shape (N, num_total_anchors, 5).
+            labels (torch.Tensor): Labels of each anchors with shape
+                (N, num_total_anchors).
+            label_weights (torch.Tensor): Label weights of each anchor with
+                shape (N, num_total_anchors)
+            bbox_targets (torch.Tensor): BBox regression targets of each anchor
+                weight shape (N, num_total_anchors, 5).
+            bbox_weights (torch.Tensor): BBox regression loss weights of each
+                anchor with shape (N, num_total_anchors, 5).
+            angle_targets (torch.Tensor): Angle classification targets of
+                each anchor weight shape (N, num_total_anchors, coding_len).
+            angle_weights (torch.Tensor): Angle classification loss weights
+                of each anchor with shape (N, num_total_anchors, 1).
+            num_total_samples (int): If sampling, num total samples equal to
+                the number of total anchors; Otherwise, it is the number of
+                positive anchors.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        # classification loss
+        labels = labels.reshape(-1)
+        label_weights = label_weights.reshape(-1)
+        cls_score = cls_score.permute(0, 2, 3,
+                                      1).reshape(-1, self.cls_out_channels)
+        loss_cls = self.loss_cls(
+            cls_score, labels, label_weights, avg_factor=num_total_samples)
+        # regression loss
+        bbox_targets = bbox_targets.reshape(-1, 5)
+        bbox_weights = bbox_weights.reshape(-1, 5)
+        # shield angle in reg. branch
+        if self.shield_reg_angle:
+            bbox_weights[:, -1] = 0.
+        bbox_pred = bbox_pred.permute(0, 2, 3, 1).reshape(-1, 5)
+        if self.reg_decoded_bbox:
+            anchors = anchors.reshape(-1, 5)
+            bbox_pred = self.bbox_coder.decode(anchors, bbox_pred)
+
+        loss_bbox = self.loss_bbox(
+            bbox_pred,
+            bbox_targets,
+            bbox_weights,
+            avg_factor=num_total_samples)
+
+        angle_cls = angle_cls.permute(0, 2, 3, 1).reshape(-1, self.coding_len)
+        angle_targets = angle_targets.reshape(-1, self.coding_len)
+        angle_weights = angle_weights.reshape(-1, 1)
+
+        loss_angle = self.loss_angle(
+            angle_cls,
+            angle_targets,
+            weight=angle_weights,
+            avg_factor=num_total_samples)
+
+        return loss_cls, loss_bbox, loss_angle
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'angle_classes'))
```

Comment:
angle_classes is not appear in `loss` function inputs.

Reply:
It should be `angle_classes` or `angle_clses`?